gh-115999: Implement thread-local bytecode and enable specialization for `BINARY_OP`

mpage opened the pull request:
This PR implements the foundational work necessary for making the specializing interpreter thread-safe in free-threaded builds and enables specialization for `BINARY_OP` as an end-to-end example. To enable future incremental work, specialization can now be toggled on a per-family basis. Subsequent PRs will enable specialization in free-threaded builds for the remaining families.

Each thread specializes a thread-local copy of the bytecode, created on the first RESUME, in free-threaded builds. All copies of the bytecode for a code object are stored in the `co_tlbc` array on the code object. Threads reserve a globally unique index identifying its copy of the bytecode in all `co_tlbc` arrays at thread creation and release the index at thread destruction. The first entry in every `co_tlbc` array always points to the "main" copy of the bytecode that is stored at the end of the code object. This ensures that no bytecode is copied for programs that do not use threads.

Thread-local bytecode can be disabled at runtime by providing either `-X tlbc=0` or `PYTHON_TLBC=0`. Disabling thread-local bytecode also disables specialization.

Concurrent modifications to the bytecode made by the specializing interpreter and instrumentation use atomics, with specialization taking care not to overwrite an instruction that was instrumented concurrently.

<!-- gh-issue-number: gh-115999 -->
* Issue: gh-115999
<!-- /gh-issue-number -->

mpage committed:
Assign threads indices into bytecode copies

mpage committed:
Get bytecode copying working

mpage committed:
Replace most usage of PyCode_CODE

mpage committed:
Refactor remove_tools

mpage committed:
Refactor remove_line_tools

mpage committed:
Instrument thread-local bytecode

mpage committed:
Use locks for instrumentation

mpage committed:
Add ifdef guards for each specialization family

mpage committed:
Specialize BINARY_OP

mpage committed:
Limit the amount of memory consumed by bytecode copies

mpage committed:
Make thread-local bytecode limits user configurable

mpage committed:
Fix a few data races when (de)instrumenting opcodes

- Fix a few places where we were not using atomics to (de)instrument
  opcodes.
- Fix a few places where we weren't using atomics to reset adaptive
  counters.
- Remove some redundant non-atomic resets of adaptive counters that
  presumably snuck as merge artifacts of https://github.com/python/cpython/pull/118064
  and https://github.com/python/cpython/pull/117144 landing close
  together.

mpage committed:
Make branch taken recording thread-safe

mpage committed:
Lock thread-local bytecode when specializing

mpage committed:
Load bytecode on RESUME_CHECK

mpage committed:
Load tlbc on generator.throw()

mpage committed:
Use tlbc instead of thread_local_bytecode

mpage committed:
Use tlbc everywhere

mpage committed:
Explicitly manage tlbc state

mpage committed:
Refactor API for fetching tlbc

mpage committed:
Add unit tests

mpage committed:
Fix initconfig in default build

mpage committed:
Fix instrumentation in default build

mpage committed:
Synchronize bytecode modifications between specialization and instrumentation using atomics

mpage committed:
Add a high-level comment

mpage committed:
Fix unused variable warning in default build

mpage committed:
Fix test_config in free-threaded builds

mpage committed:
Fix formatting

mpage committed:
Remove comment

mpage committed:
Fix data race in _PyInstruction_GetLength

Read the opcode atomically, the interpreter may be specializing it

mpage committed:
Fix tier2 optimizer

uops that uses `this_instr` are not eligible to be included in traces.
Restore `RESUME_CHECK` to its original implemenation and inline reloading
bytecode, which does not need to reset `this` since nothing else in
`RESUME_CHECK` uses it.

mpage committed:
Use __VA_ARGS__ for macros

Named argument lists apparently aren't supported on windows

mpage committed:
Update vcxproj files to include newly added files

Hopefully I did this correctly.

mpage committed:
Mark unused params

mpage committed:
Keep tier2 and the JIT disabled in free-threaded builds

mpage committed:
Only allow enabling/disabling tlbc

mpage committed:
Update libpython for gdb

mpage committed:
Merge branch 'main' into gh-115999-thread-local-bytecode

mpage committed:
Handle out of memory errors

mpage committed:
Merge branch 'main' into gh-115999-thread-local-bytecode

mpage committed:
Fix warnings on windows

mpage committed:
Fix another warning

mpage committed:
Ugh actually fix it

corona10 commented on Objects/codeobject.c:
```diff
@@ -2633,5 +2675,129 @@ _PyCode_Fini(PyInterpreterState *interp)
+    // initial must be a power of two
+    assert(!(initial & (initial - 1)));
+    Py_ssize_t res = initial;
+    while (res && res < limit) {
```
Nit: 
How about leaving a comment as `TODO` about investigating the possibility of a bit of twiddling hack here?

mpage replied:
I like the idea, but I don't think this will be performance sensitive enough to warrant a more complex implementation. This happens in the slow path when the array needs to be resized, which should happen pretty rarely, and the overall cost will likely be dominated by the locking and allocation overhead. Plus, we can always tackle this later if it shows up in a profile as a hot spot.

Fidget-Spinner commented on Python/index_pool.c:
```diff
@@ -0,0 +1,193 @@
+#include <stdbool.h>
```
Could you please document somewhere (either the header file or this file) what this is for?

mpage replied:
Happy to do that but I'm not sure it's worth it. It provides `bool`. If this is confusing, I can replace the lone use of `bool` with `int` and remove the include.

JelleZijlstra replied:
I think the question is about the purpose of this file as a whole, not the specific #include.

Fidget-Spinner replied:
Yeah sorry I wasn't clear. I meant the whole file. I know what `stdbool` does haha.

mpage replied:
lol that makes a lot more sense. Yeah, of course!

markshannon commented on Lib/test/test_tlbc.py:
```diff
@@ -0,0 +1,197 @@
+"""Tests for thread-local bytecode."""
```
Can you rename the file to something like `test_thread_local_bytecode`?
`tlbc` is a bit obscure

mpage committed:
Add high-level comment about index pools

mpage committed:
Merge branch 'main' into gh-115999-thread-local-bytecode

mpage committed:
Merge branch 'main' into gh-115999-thread-local-bytecode

mpage committed:
Exclude tlbc from refleak counts

TLBC is intended to be reused across threads with different lifetimes,
so may appear as a "leak" depending on the order in which threads execute
code objects. They are freed when the code object is destroyed, which
typically occurs when the runtime is finalized.

mpage committed:
Merge branch 'main' into gh-115999-thread-local-bytecode

mpage committed:
Regen files

mpage committed:
Move `get_tlbc_blocks` into the sys module

mpage committed:
Merge branch 'main' into gh-115999-thread-local-bytecode

corona10 reviewed (approved):
Overall architecture looks good to me.
But I will delegate the merging of this PR to @Fidget-Spinner with my +1.

Great job @mpage

mpage committed:
Merge branch 'main' into gh-115999-thread-local-bytecode

Fidget-Spinner commented on Objects/codeobject.c:
```diff
@@ -2633,5 +2675,129 @@ _PyCode_Fini(PyInterpreterState *interp)
+        tlbc = new_tlbc;
+    }
+    char *bc = PyMem_Calloc(1, _PyCode_NBYTES(co));
+    if (bc == NULL) {
```
Just checking, this doesn't leak memory by not freeing `new_tlbc` because that's already owned by the code object right?

mpage replied:
Yes, that’s correct.

Fidget-Spinner reviewed (approved):
Just one question.

markshannon commented on Python/specialize.c:
```diff
@@ -438,13 +457,12 @@ do { \
 // Initialize warmup counters and insert superinstructions. This cannot fail.
 void
-_PyCode_Quicken(PyCodeObject *code)
+_PyCode_Quicken(_Py_CODEUNIT *instructions, Py_ssize_t size)
```
Why change this?

mpage replied:
So that we can quicken copies of the bytecode as they are created.

markshannon replied:
Can we quicken once, and copy as needed?

I would like `_PyCode_Quicken` to be called only once as part of code object creation. That way we can integrate it into the process of loading pyc files in the future.

markshannon commented on Include/internal/pycore_code.h:
```diff
@@ -600,6 +631,36 @@ struct _PyCode8 _PyCode_DEF(8);
+// Return a pointer to the thread-local bytecode for the current thread, if it
+// exists.
+static inline _Py_CODEUNIT *
+_PyCode_GetTLBCFast(PyCodeObject *co)
```
This function seems to be exposed just for testing. It is only used in one place AFAICT.
Maybe inline it into `_PyEval_GetExecutableCode` and test that?

mpage replied:
A few things:

1. This is just a personal preference, but I'd rather keep the implementation details contained in the code object files and rely on the compiler to handle the fast path inlining for us.
2. In tests we want to retrieve the thread-local bytecode only if it exists, to check that a copy has been created. `_PyEval_GetExecutableCode` will create the copy if it does not exist.

markshannon commented on Include/internal/pycore_code.h:
```diff
@@ -600,6 +631,36 @@ struct _PyCode8 _PyCode_DEF(8);
+_PyCode_GetTLBCFast(PyCodeObject *co)
+{
+    _PyCodeArray *code = _Py_atomic_load_ptr_acquire(&co->co_tlbc);
+    _PyThreadStateImpl *tstate = (_PyThreadStateImpl *) PyThreadState_GET();
```
This is slow. The threadstate is readily available to all callers, so pass it as a parameter.

markshannon commented on Include/cpython/code.h:
```diff
@@ -72,6 +72,24 @@ typedef struct {
+ */
+typedef struct {
+    Py_ssize_t size;
+    char *entries[1];
```
```suggestion
    _Py_CODEUNIT *entries[1];
```

mpage replied:
I started out with this, but it would require including `pycore_code.h`, which introduces a circular dependency between the two files. That should probably be sorted out in a separate PR. I'm guessing that's also why `co_code_adaptive` is a `char` array and not a `_Py_CODEUNIT` array:

https://github.com/python/cpython/blob/99400930ac1d4e5e10a5ae30f8202d8bc2661e39/Include/cpython/code.h#L140

markshannon replied:
That makes sense. We should probable move `PyCodeObject` to an internal header sometime in the future.

markshannon commented on Python/bytecodes.c:
```diff
@@ -190,7 +190,24 @@ dummy_func(
+                next_instr = frame->instr_ptr + 1;
+            }
+            #else
+            (void)this_instr;
```
Not necessary

markshannon commented on Python/bytecodes.c:
```diff
@@ -190,7 +190,24 @@ dummy_func(
+                int off = this_instr - frame->bytecode;
+                frame->bytecode = bytecode;
+                frame->instr_ptr = frame->bytecode + off;
+                this_instr = frame->instr_ptr;
```
`this_instr` is generated. Do not assign to it.
https://github.com/python/cpython/issues/125039

mpage replied:
We need to do this so that `this_instr` is correct in any uops that follow. Concretely, in `_MAYBE_INSTRUMENT`.

markshannon replied:
This is not how it works, `this_instr` is a pointer to the current instruction. You can read it, but not write it.
The code generator is free to create and rewrite it as it wants.

If you need to change `next_instr` for following code it will need to be its own instruction, like a jump.

mpage replied:
> This is not how it works, this_instr is a pointer to the current instruction.

I understand that, but `this_instr` will be incorrect after we reload the bytecode. There may be uops that follow `_LOAD_BYTECODE` that use `this_instr` (e.g. `_MAYBE_INSTRUMENT`) and we want them to use the correct value. 

IIRC, right now the code generator sets `this_instr` once at the beginning of the generated opcode body if any of the uops reference `this_instr`. I could modify the code generator to recognize `_LOAD_BYTECODE` and generate code to reload `this_instr`  after it as necessary. I don't like the special casing, but maybe that's slightly better than what's here. What do you think?

mpage replied:
I ended up working around this by dispatching to the same instruction when we need to reload the bytecode. It'll make the first function call a tad bit slower, but we can figure out a better solution if needed.

markshannon commented on Python/bytecodes.c:
```diff
@@ -204,6 +221,19 @@ dummy_func(
             uintptr_t version = FT_ATOMIC_LOAD_UINTPTR_ACQUIRE(_PyFrame_GetCode(frame)->_co_instrumentation_version);
             assert((version & _PY_EVAL_EVENTS_MASK) == 0);
             DEOPT_IF(eval_breaker != version);
+            #ifdef Py_GIL_DISABLED
```
`RESUME_CHECK` should be what it says, a check. It should only perform fast checks and leave the heavy lifting to `RESUME`.

mpage replied:
I'll have this call `_PyCode_GetTLBCFast` (only a few loads) and deopt to `RESUME` if a copy of the bytecode needs to be created.

markshannon replied:
What about storing the thread's tlbc index in the frame. It is much smaller than a full pointer, and this check remains cheap(ish) `frame->tlbc_index != tstate->tlbc_index`.


Fetching the bytecode shouldn't be too expensive either: `bytecode = code->tlbc->entries[frame->tlbc_index]`

mpage replied:
I'd prefer not to add anything to the frame. Do you think it's too expensive to iterate over the bytecode copies when we need to compute the instruction offset? If that's a no-go, then storing `tlbc_index` on the frame sounds good to me.

mpage replied:
Ended up going with storing `tlbc_index` on the frame. Thanks for the suggestion!

markshannon commented on Lib/test/libregrtest/refleak.py:
```diff
@@ -145,6 +145,11 @@ def get_pooled_int(value):
             # Use an internal-only keyword argument that mypy doesn't know yet
             _only_immortal=True)  # type: ignore[call-arg]
         alloc_after = getallocatedblocks() - interned_immortal_after
+        if _get_tlbc_blocks := getattr(sys, "_get_tlbc_blocks", None):
```
We are still allocating these blocks.
Alter the tests to account for their presence, but I don't think we should discount them.
We still want to check that they are released at shutdown.

mpage replied:
I don't think it will be practical to modify the tests so that they pass reliably in regrtest mode and it will create a confusing experience for people who add new tests. The issue is that new copies of the bytecode are created the first time a thread calls a function. In tests that use multiple threads it's possible that a different thread ends up calling some function each time the test is run. This results in a new copy of the bytecode being created for each test run, which regrtest treats as a leak.

More concretely, imagine we have a test using four threads, and regrtest is running 3 warmup iterations and 3 test iterations, and we encounter the following sequence:

1. Function F is called by thread 0 in the 3 warmup iterations. This will create one copy of the bytecode.
2. F is called by thread 1 in the first test iteration. This will create one copy of the bytecode. regrtest will treat this as a leaked object.
3. F is called by thread 2 in the second test iteration. This creates another copy of the bytecode that is treated as a leak.
4. F is called by thread 3 in the third test iteration.  This creates another copy of the bytecode that is treated as a leak.

To make this pass reliably we'd need to either that all of threads create their copies of the bytecode in the warmup runs.

Since what we really care about is checking that bytecode copies are cleaned up, I can write an isolated test for that.

markshannon replied:
We have similar issues with the JIT, which allocates blocks at hard-to-predict times.
If a test is using multiple threads, then they will have all run in the first iteration of the test. The refleak tests do not spawn additional threads (I think).

How many tests are failing without this?

mpage replied:
> If a test is using multiple threads, then they will have all run in the first iteration of the test. The refleak tests do not spawn additional threads (I think).

The issue is not related to thread creation, it's related to when threads call functions for the first time (which requires creating a new copy of the bytecode). If a function is only called once during the duration of a test (e.g. `notify` on a condition variable), then it's possible that a different thread will end up calling the function each time the test is run. This will appear as a leak in the refleak tests, since each run of the test in this scenario will result in a new copy of the bytecode being created.

> How many tests are failing without this?

It's a relatively small number of test suites (`test_threading`, `test_queue` and `test_io`), but they fail intermittently.

JelleZijlstra replied:
Do we not clean up the copy of the bytecode when the thread exits?

mpage replied:
> Do we not clean up the copy of the bytecode when the thread exits?

No, they remain in the array attached to the code object for reuse by the next thread that is assigned the same tlbc index.

markshannon replied:
I still think we need to count the blocks, otherwise we cannot spot regressions. 

Can we modify the failing tests somehow to be a bit more predictable in how many threads they use?
Or increase the number of iterations to stabilize those tests?

mpage replied:
> Can we modify the failing tests somehow to be a bit more predictable in how many threads they use?

The issue isn't the number threads that a test uses, per-se. It's whether or not a thread calls a function for the first time during the execution of a test.

> Or increase the number of iterations to stabilize those tests?

This might be possible, but there's no guarantee that it would work, and it would require increasing the number of iterations for *all* tests that are run during refleak tests. AFAIK there's no way to manually specify the number of iterations required for a specific test. (This could be added, of course.)

> I still think we need to count the blocks, otherwise we cannot spot regressions.

Agreed that we want to catch regressions. I added a [test](https://github.com/python/cpython/pull/123926/files#diff-01d18615b7d1791994f2cd69f42a07b9c9fdb07b7c0081477ec925e12e2b3701R195-R230) to verify that copies of the bytecode are cleaned up when code objects are destroyed.

markshannon commented on Include/internal/pycore_code.h:
```diff
@@ -313,11 +314,41 @@ extern int _PyLineTable_PreviousAddressRange(PyCodeAddressRange *range);
 #ifdef Py_GIL_DISABLED
 // gh-115999 tracks progress on addressing this.
-#define ENABLE_SPECIALIZATION 0
+#define ENABLE_SPECIALIZED_BINARY_OP ENABLE_SPECIALIZATION
```
We don't need 20 odd different versions of `ENABLE_SPECIALIZED`.
I think we only need two: `ENABLE_SPECIALIZED` and `ENABLE_SPECIALIZED_FT`

Change the use of `ENABLE_SPECIALIZED` in `BINARY_OP` to `ENABLE_SPECIALIZED_FT`.
```C
#define ENABLE_SPECIALIZED_FT defined(Py_GIL_DISABLED) && ENABLE_SPECIALIZATION
```

markshannon commented on Python/bytecodes.c:
```diff
@@ -4698,15 +4719,15 @@ dummy_func(
             _PyExitData *exit = (_PyExitData *)exit_p;
             PyCodeObject *code = _PyFrame_GetCode(frame);
-            _Py_CODEUNIT *target = _PyCode_CODE(code) + exit->target;
+            _Py_CODEUNIT *target = _PyFrame_GetBytecode(frame) + exit->target;
```
~We execute `_EXIT_TRACE` a lot, and the extra overhead is going to be significant.~
I was thinking of the tlbc functions. This is cheap, so no problem.

markshannon commented on Python/index_pool.c:
```diff
@@ -0,0 +1,193 @@
+#include <stdbool.h>
+
```
What's the advantage of using a heap over a simple freelist?
It seems slower and more complicated.

Once gaps appear in the array of indexes, does it matter if they are filled from the bottom or not?
If it does, I suspect that scanning a freelist would be faster anyway due to better locality of reference.

mpage replied:
We want to assign the smallest available id to newly created threads, because it impacts the size of the arrays  attached to each code object that store pointers to thread-local bytecode. This is contrived, but imagine this scenario:

1. Create 100 threads.
2. Destroy the threads from (1).
3. Create a new thread.

In step (3) we have 100 indices to choose from. If we pick randomly and pick 99, then each code object that is called by the thread from (3) will end up with an array of size 100 attached to it. By picking the smallest available index we can minimize the size of the bytecode array.

I don't anticipate this being much of a problem in practice, but it seems worth being mindful of it, especially when it's not particularly complicated to do so.

As for a heap vs a free-list, I don't think the heap adds that much complexity, and has better asymptotics for finding the minimum. It's also implemented on top of an array, so I'm not sure how much weight to give to the locality of reference argument.

markshannon commented on Objects/codeobject.c:
```diff
@@ -6,16 +6,20 @@
 #include "pycore_tuple.h"         // _PyTuple_ITEMS()
 #include "clinic/codeobject.c.h"
 
+#define INITIAL_SPECIALIZED_CODE_SIZE 16
```
OOI, why 16?

mpage replied:
This was pretty arbitrary. I wanted to pick something that was large enough to not require too many resizes to reach a steady state, but also small enough to minimize wasted memory in the cases where a small number of threads were used.

markshannon commented on Python/bytecodes.c:
```diff
@@ -190,7 +190,24 @@ dummy_func(
+            #ifdef Py_GIL_DISABLED
+            _Py_CODEUNIT *bytecode = _PyEval_GetExecutableCode(_PyFrame_GetCode(frame));
+            ERROR_IF(bytecode == NULL, error);
+            if (frame->bytecode != bytecode) {
```
Since a reload of the bytecode is only necessary when the threads differ, would it make sense to store a pointer to the currently executing thread on the frame, instead of a pointer to the bytecode?

We only need the `bytecode` pointer to compute offsets, which only happens in debugging and instrumentation code.

mpage replied:
I'm not sure that helps simplify things. We always have to call `_PyEval_GetExecutableCode` in `RESUME`, even if the threads are the same, because it may be the first time that the thread is executing the code object (and it will need to create its copy of the bytecode).

I think we can get rid of the `bytecode` pointer if we don't care too much about the performance of `PyFrame_GetBytecode`. We could compute the `bytecode` pointer on demand by iterating through the copies of the thread-local bytecode and seeing which one contained the `instr_ptr` for the frame. This would be O(N), but maybe that's not a problem if we don't care much about perf and N is relatively small.

markshannon replied:
I don't think we need to iterate over the array: `bytecodes = code->co_tlbc->entries[tstate->tlbc_index]`

mpage replied:
That doesn't work if the thread that needs to compute `bytecodes` is not the thread that is executing the frame (this happens in instrumentation). We need to retrieve the pointer to the bytecode that contains the frame's `instr_ptr` in order to compute the offset correctly. To do that we either need something on the frame (e.g. the `bytecode` pointer or the `tlbc_index` as you suggested earlier) or to loop through the copies of the bytecode for the code object and check which one contains `instr_ptr`.

markshannon reviewed (changes requested):
I've some comments and suggestions.

Is there any limit to the number of copies of the bytecode?
If we have a 1000 threads could we get a 1000 copies of the bytecode?

mpage committed:
Merge branch 'main' into gh-115999-thread-local-bytecode

mpage committed:
Work around `this_instr` now being const

mpage committed:
Make RESUME_CHECK cheaper

mpage committed:
Pass tstate to _PyCode_GetTLBCFast

mpage committed:
Rename test_tlbc.py to test_thread_local_bytecode.py

mpage committed:
Remove per-family defines for specialization

mpage committed:
Replace bytecode pointer with tlbc_index

This uses less space

mpage committed:
Add a test verifying that we clean up tlbc when the code object is destroyed

mpage committed:
Merge branch 'main' into gh-115999-thread-local-bytecode

mpage committed:
Fix indentation

mpage committed:
Clarify comment

mpage committed:
Fix TSAN

mpage committed:
Add test for cleaning up tlbc in correct place, not old emacs buffer

mpage committed:
Remove test_tlbc.py

markshannon commented on Include/internal/pycore_frame.h:
```diff
@@ -68,6 +68,10 @@ typedef struct _PyInterpreterFrame {
     _Py_CODEUNIT *instr_ptr; /* Instruction currently executing (or about to begin) */
+#ifdef Py_GIL_DISABLED
+    /* Index of thread-local bytecode containing instr_ptr. */
+    Py_ssize_t tlbc_index;
```
Why signed? I'm guessing that's so it can be set to a negative value to indicate "invalid"?
I think 2 billion copies should be sufficient, so `int32_t` will do. It doesn't make any difference now due to alignment, but I'd like to keep interpreter frames as small as possible.

mpage replied:
> Why signed? I'm guessing that's so it can be set to a negative value to indicate "invalid"?

Yep, exactly. Updated to use `int32_t`.

markshannon commented on Include/internal/pycore_frame.h:
```diff
@@ -76,14 +80,27 @@ typedef struct _PyInterpreterFrame {
 }
 
+static inline _Py_CODEUNIT *
+_PyFrame_GetBytecode(_PyInterpreterFrame *f)
```
You were storing the bytecode in the frame directly before, IIRC.
This looks more expensive, and is used on at least one fast path:
https://github.com/python/cpython/pull/123926/files#diff-729a985b0cb8b431cb291f1edb561bbbfea22e3f8c262451cd83328a0936a342R4821

Does it makes things faster overall, or is it just more compact?

mpage replied:
> You were storing the bytecode in the frame directly before, IIRC.

Yep. You suggested storing `tlbc_index` instead since it was smaller. I think this is better for a couple of reasons:

1. It's smaller, as you said.
2. It simplifies and speeds up the implementation of `RESUME_CHECK`. Previously, we would have to load the bytecode pointer for the current thread and deopt if it didn't match what was in the frame. Now we only have to compare tlbc indices. This is a cost shift, however, since now the callers of `_PyFrame_GetBytecode` have to do the more expensive load of the bytecode. I think the size reduction + simplification of `RESUME_CHECK` probably outweighs the higher cost of `_PyFrame_GetBytecode`. tier2 is also still disabled in free-threaded builds, so it's a bit hard to evaluate the relative cost of slower trace exits vs faster `RESUME_CHECK`s. Once we get tier2 enabled we can reevaluate.

markshannon commented on Objects/typeobject.c:
```diff
@@ -11617,8 +11618,8 @@ super_init_without_args(_PyInterpreterFrame *cframe, PyCodeObject *co,
             // to use _PyOpcode_Deopt here:
-            assert(_PyCode_CODE(co)[0].op.code == MAKE_CELL ||
-                   _PyCode_CODE(co)[0].op.code == COPY_FREE_VARS);
+            assert(_PyFrame_GetBytecode(cframe)[0].op.code == MAKE_CELL ||
```
`MAKE_CELL` and `COPY_FREE_VARS` are never specialized, so we can read any bytecode, including the default one.
I think `_PyCode_CODE(co)` is fine.

markshannon commented on Python/ceval_macros.h:
```diff
@@ -318,6 +310,17 @@ GETITEM(PyObject *v, Py_ssize_t i) {
+
+#ifdef ENABLE_SPECIALIZATION_FT
+/* Multiple threads may execute these concurrently if the thread-local bytecode
+ * limit is reached and they all execute the main copy of the bytecode. This is
```
This comment is out of date.

markshannon commented on Python/specialize.c:
```diff
@@ -466,9 +484,28 @@ _PyCode_Quicken(PyCodeObject *code)
+        int caches = _PyOpcode_Caches[opcode];
+        if (caches) {
+            instructions[i + 1].counter =
+                initial_unreachable_backoff_counter();
```
This is the wrong value for branches. It probably doesn't matter what you set them to, but it might as well be 0x5555;

markshannon commented on Python/specialize.c:
```diff
@@ -466,9 +484,28 @@ _PyCode_Quicken(PyCodeObject *code)
+#ifdef Py_GIL_DISABLED
+
+void
+_PyCode_DisableSpecialization(_Py_CODEUNIT *instructions, Py_ssize_t size)
```
I'm not sure that an extra function is necessary.

The only difference between "quicken" and "disable" is the initial counter values.

If you want, we could pass that in as an extra parameter to `_PyCode_Quicken`.

So `_PyCode_Quicken(insts, size)` would become `_PyCode_Quicken(insts, size, 1)`
and `_PyCode_DisableSpecialization` would become 
`_PyCode_Quicken(insts, size, 0)`

Feel free to rename `_PyCode_Quicken` to something like `_PyCode_SetCounters` as it doesn't do any quickening anymore.

markshannon commented on Include/internal/pycore_frame.h:
```diff
@@ -144,13 +161,33 @@ static inline void _PyFrame_Copy(_PyInterpreterFrame *src, _PyInterpreterFrame *
 static inline void
 _PyFrame_Initialize(
-    _PyInterpreterFrame *frame, _PyStackRef func,
+    PyThreadState *tstate, _PyInterpreterFrame *frame, _PyStackRef func,
```
The only purpose of passing the thread state is to initialize the tlbc index, IIUC.
Most callers will already have a tlbc index, so could we pass that instead?

mpage replied:
> Most callers will already have a tlbc index, so could we pass that instead?

The tlbc index is only present in free-threaded builds. To do this I think we'd need to either have separate versions of `_PyFrame_Initialize` for free-threaded/default builds or introduce the notion of thread-local bytecode into the default build. Passing the `PyThreadState` as the first parameter seems simpler and more maintainable.

markshannon reviewed (commented):
Definitely getting close to being mergeable.

A few more comments and questions though.

mpage committed:
Merge branch 'main' into gh-115999-thread-local-bytecode

mpage committed:
Use int32_t instead of Py_ssize_t for tlbc indices

mpage committed:
Use _PyCode_CODE instead of PyFrame_GetBytecode in super_init_without_args

mpage committed:
Update comment

mpage committed:
Consolidate _PyCode_{Quicken,DisableSpecialization} into _PyCode_InitCounters

Yhg1s commented on Include/cpython/code.h:
```diff
@@ -138,6 +156,7 @@ typedef struct {
        Type is a void* to keep the format private in codeobject.c to force     \
        people to go through the proper APIs. */                                \
     void *co_extra;                                                            \
+    _PyCode_DEF_THREAD_LOCAL_BYTECODE()                                        \
```
Is there a reason not to put this at the end of the struct (after co_code_adaptive[]) so that the layout of the struct doesn't change between build modes?

mpage replied:
That is possible, but it would mean that we cannot fetch the tlbc array from a fixed offset. The code object is laid out such that there is a fixed-length part (what we see in `_PyCode_DEF`) followed by the variable-sized bytecode. `co_code_adaptive` is a flexible array member that points to the start of the variable-sized bytecode. In order to not change the struct layout we would need to put the tlbc array pointer after the variable-sized bytecode. Fetching it would require dynamically computing its location. I'd prefer not to do that, and instead have it live at a fixed offset, because we need to fetch it on every function call.

Yhg1s replied:
Ah, sorry, I hadn't noticed `SIZE` was parameterised, sorry! Yes, that make sense.

Yhg1s commented on Include/internal/pycore_frame.h:
```diff
@@ -68,6 +68,10 @@ typedef struct _PyInterpreterFrame {
     PyObject *f_locals; /* Strong reference, may be NULL. Only valid if not on C stack */
     PyFrameObject *frame_obj; /* Strong reference, may be NULL. Only valid if not on C stack */
     _Py_CODEUNIT *instr_ptr; /* Instruction currently executing (or about to begin) */
+#ifdef Py_GIL_DISABLED
```
Similar here: can we move it to the end of the struct to keep the layout the same between build modes? (I expect this will make life easier for debuggers/profilers that care about entries after instr_ptr.)

mpage replied:
I'd prefer not to do that for pretty much that same reasons as above. In this case the locals and stack follow the fixed part of `_PyInterpreterFrame`. We need to fetch the `tlbc_index` on every function call, so I'd prefer to keep it at a fixed location to minimize overhead.

Yhg1s commented on Lib/test/test_cmd_line.py:
```diff
@@ -1068,6 +1069,40 @@ def res2int(self, res):
+            t.start()
+            t.join()"""
+        assert_python_ok("-W", "always", "-X", "tlbc=0", "-c", code)
+        assert_python_ok("-W", "always", "-c", code, PYTHON_TLBC="0")
```
Maybe add a test that verifies PYTHON_TLBC=1 also works, and add a comment that TLBC, including the -X argument, is more extensively covered in its own tests. (My initial reaction here was to ask for a test covering `-X tlbc=1` too.)

Yhg1s commented on Python/ceval_macros.h:
```diff
@@ -318,6 +310,18 @@ GETITEM(PyObject *v, Py_ssize_t i) {
+    FT_ATOMIC_STORE_UINT16_RELAXED(       \
+        bitset, (FT_ATOMIC_LOAD_UINT16_RELAXED(bitset) << 1) | (flag))
+#else
+#define RECORD_BRANCH_TAKEN(bitset, flag)
```
Shouldn't the macro expand to the old, non-atomic code in a build without ENABLE_SPECIALIZATION_FT? (Or, why have that macro at all, since it's always set to 1?)

mpage replied:
> Or, why have that macro at all, since it's always set to 1?

It's a temporary measure to allow us to gradually enable specialization in free-threaded builds by replacing uses of `ENABLE_SPECIALIZATION` (set to `0` in free-threaded builds) with `ENABLE_SPECIALIZATION_FT` (set to `1` in free-threaded builds and `ENABLE_SPECIALIZATION` in default builds). Eventually we'll set `ENABLE_SPECIALIZATION` to `1` in free-threaded builds too and replace all of the uses `ENABLE_SPECIALIZATION_FT` with `ENABLE_SPECIALIZATION`.

Yhg1s replied:
I still don't quite get why this should expand to empty code if it's not set, but I guess it doesn't matter if it's a temporary measure anyway :)

Yhg1s commented on Python/index_pool.c:
```diff
@@ -0,0 +1,193 @@
+}
+
+static inline Py_ssize_t
+left_child(int32_t i)
```
Why are these int32_t instead of Py_ssize_t? (See compiler warnings from MSVC++.)

mpage replied:
Oops, those should be `Py_ssize_t`. Good catch!

mpage committed:
Merge branch 'main' into gh-115999-thread-local-bytecode

mpage committed:
Fix incorrect types

mpage committed:
Add command-line tests for enabling TLBC

mpage committed:
Update libpython.py for tlbc_index

mpage committed:
Avoid special casing in _PyEval_GetExecutableCode

It's cleaner to assign all threads the index of the main copy of the
bytecode when tlbc is disabled rather than adding a special case
in _PyEval_GetExecutableCode.

mpage committed:
Merge branch 'main' into gh-115999-thread-local-bytecode

mpage committed:
Merge branch 'main' into gh-115999-thread-local-bytecode

mpage committed:
Clear TLBC when other caches are cleared

mpage committed:
Remove _get_tlbc_blocks

Yhg1s commented on Lib/test/test_sys.py:
```diff
@@ -1080,7 +1080,14 @@ def test_getallocatedblocks(self):
+            # references, this can cause the total number of allocated
+            # blocks to exceed the total number of references.
+            if not support.Py_GIL_DISABLED:
+                self.assertLess(a, sys.gettotalrefcount())
```
Does this still report leaks even when clearing unused TLBC? I would expect regrtest to make sure no other threads are running at this point, for this test. (If we do make regrtest multithreaded I expect this test will need to be special-cased anyway.)

mpage replied:
This is problematic when run on its own (i.e. `./python -m test test_sys`). It seems to be very sensitive to which kinds of objects are on the heap as well as the number of non reference counted allocations (blocks) per object. With the introduction of TLBC, there is at least one additional block allocated per code object that is not reference counted (the `_PyCodeArray`). That is enough to trigger the assertion.

This assertion feels pretty brittle and I'd be in favor of removing it, but that's probably worth doing in a separate PR.

markshannon commented on Lib/test/test_sys.py:
```diff
@@ -1080,7 +1080,14 @@ def test_getallocatedblocks(self):
+            # code objects is a large fraction of the total number of
+            # references, this can cause the total number of allocated
+            # blocks to exceed the total number of references.
+            if not support.Py_GIL_DISABLED:
```
Now that we can free the unused tlbcs, can we replace this with `sys._clear_internal_caches()`?

mpage replied:
Unfortunately, no. It seems to be very sensitive to which kinds of objects are on the heap as well as the number of non reference counted allocations (blocks) per object. With the introduction of TLBC there is at least one additional block allocated per code object that is not reference counted, the _PyCodeArray, which is present even if we free the unused TLBCs. Its presence is enough to trigger the assertion.

This assertion feels pretty brittle and I'd be in favor of removing it, but that's probably worth doing in a separate PR.

markshannon replied:
Maybe replace it with a more meaningful test rather than remove it. But in another PR.

markshannon reviewed (commented):
Looks good.

One question. Can we prefix the test for leaking blocks with `sys._clear_internal_caches()` instead of making it conditional on not using free-threading?

markshannon reviewed (approved):
I still have concerns about memory use, but we can iterate on that in subsequent PRs.

mpage committed:
Merge branch 'main' into gh-115999-thread-local-bytecode

mpage committed:
Rename _PyCode_InitCounters back to _PyCode_Quicken

We are quickening LOAD_CONST now.

mpage committed:
Merge branch 'main' into gh-115999-thread-local-bytecode

mpage committed:
Merge branch 'main' into gh-115999-thread-local-bytecode
